tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.partitioner == partitioner {
+		return
+	}
+
+	t.partitioner = partitioner
+
+	err := t.resetRing()
+	if err != nil {
+		Logger.Println(err)
+	}
+}
+
+// resetRing replaces the token ring with a new one.
+// must be called with t.mu lock held for writing.
+func (t *TokenAwareBanditHostPolicy) resetRing() error {
+	newTR, err := newTokenRing(t.partitioner, t.hostInfoSlice())
+	if err != nil {
+		return err
+	}
+	t.tokenRing = newTR
+
+	for keyspaceName := range t.keyspaces {
+		t.updateReplicas(keyspaceName)
+	}
+
+	return nil
+}
+
+// AddHost adds a single host to the policy after it was discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHost(host *HostInfo) {
+	t.AddHosts([]*HostInfo{host})
+}
+
+// AddHosts adds multiple hosts to the policy after they were discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHosts(hosts []*HostInfo) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	added := false
+
+	for _, host := range hosts {
+		hostID := host.HostID()
+		if hostID == "" {
+			continue
+		}
+		_, ok := t.hosts[hostID]
+		if !ok {
+			t.hosts[hostID] = &tokenAwareBanditHost{
+				hostInfo: host,
+			}
+			added = true
+		}
+	}
+
+	if !added {
+		return
+	}
+
+	err := t.resetRing()
+	if err != nil {
+		t.logger.Println(err)
+	}
+}
+
+// RemoveHost from the policy after it was removed from the cluster.
+func (t *TokenAwareBanditHostPolicy) RemoveHost(host *HostInfo) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	hostID := host.HostID()
+	if hostID == "" {
+		return
+	}
+
+	if _, ok := t.hosts[hostID]; !ok {
+		return
+	}
+
+	delete(t.hosts, hostID)
+	err := t.resetRing()
+	if err != nil {
+		t.logger.Println(err)
+	}
+}
+
+func (t *TokenAwareBanditHostPolicy) HostUp(host *HostInfo) {}
+
+func (t *TokenAwareBanditHostPolicy) HostDown(host *HostInfo) {}
+
+// findReplicas gets replicas for the query.
+// If for any reason we can't find replicas, the returned slice's length will be zero.
+// t.mu must be held at least read-only.
+func (t *TokenAwareBanditHostPolicy) findReplicas(qry ExecutableQuery) (token, []*HostInfo) {
+	if qry == nil {
+		return nil, nil
+	}
+
+	routingKey, err := qry.GetRoutingKey()
+	if err != nil || routingKey == nil {
+		return nil, nil
+	}
+
+	if t.tokenRing == nil {
+		return nil, nil
+	}
+
+	partitioner := qry.GetCustomPartitioner()
+	if partitioner == nil {
+		partitioner = t.tokenRing.partitioner
+	}
+
+	token := partitioner.Hash(routingKey)
+
+	var ringReplicas tokenRingReplicas
+	if keyspace, ok := t.keyspaces[qry.Keyspace()]; ok {
+		ringReplicas = keyspace.replicas
+	}
+
+	// it's safe to call replicasFor on nil value.
+	ht := ringReplicas.replicasFor(token)
+
+	var replicas []*HostInfo
+	if ht == nil {
+		// If we don't have information about replicas yet, fall back to primary replica for the token.
+		// This can happen before we know the topology strategy for a keyspace.
+		host, _ := t.tokenRing.GetHostForToken(token)
```

Comment:
What are we ignoring here? Is it possible that `host` will be `nil` and we will return `[]*HostInfo{nil}` on the next line? Won't that cause issues further down the code?

Reply:
`func (t *tokenRing) GetHostForToken(token token) (host *HostInfo, endToken token) {`

`endToken` is the token assigned to the node that we find by token. It was used for lookups before, but is not used anymore (except in tests). This was changed in https://github.com/gocql/gocql/pull/1350. We can probably remove endToken altogether.